src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs

```diff
@@ -4236,5 +4236,94 @@ static void M(Span<int> s)
             var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
             comp.VerifyDiagnostics();
         }
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void SingleOverloadReadOnlySpan()
+        {
+            string source = """
+                using System;
+                
+                ReadOnlySpan<char> s = "0123";
+                Console.Write(s[1..].ToString());
+                """;
+            var comp = CompileAndVerify(source, targetFramework: TargetFramework.Net70, expectedOutput: "123");
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("<top-level-statements-entry-point>", """
+            {
+              // Code size       39 (0x27)
+              .maxstack  2
+              .locals init (System.ReadOnlySpan<char> V_0, //s
+                            System.ReadOnlySpan<char> V_1)
+              IL_0000:  ldstr      "0123"
+              IL_0005:  call       "System.ReadOnlySpan<char> string.op_Implicit(string)"
+              IL_000a:  stloc.0
+              IL_000b:  ldloca.s   V_0
+              IL_000d:  ldc.i4.1
+              IL_000e:  call       "System.ReadOnlySpan<char> System.ReadOnlySpan<char>.Slice(int)"
+              IL_0013:  stloc.1
+              IL_0014:  ldloca.s   V_1
+              IL_0016:  constrained. "System.ReadOnlySpan<char>"
+              IL_001c:  callvirt   "string object.ToString()"
+              IL_0021:  call       "void System.Console.Write(string)"
+              IL_0026:  ret
+            }
+            """);
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void SingleOverloadSpan()
+        {
+            string source = """
+                using System;
+                
+                Console.Write("0123".ToCharArray().AsSpan()[1..].ToString());
```

Comment:
Consider separating out the actual indexing into a local function, so that we can make the IL assert as minimal as possible.

Reply:
It looks like impossible.
Do you know how to fix?

```cs
[Fact]
public void SingleOverloadSpan()
{
    if (!ExecutionConditionUtil.IsCoreClr)
        return;

    string source = """
        using System;
        
        Console.Write(SliceSpan("0123".ToCharArray().AsSpan()).ToString());

        Span<char> SliceSpan(Span<char> s) => s[1..];
        """;
    var comp = CompileAndVerify(source, targetFramework: TargetFramework.Net70, expectedOutput: "123");
    comp.VerifyDiagnostics();
    // ↓ I'll fix it later
    comp.VerifyIL("<top-level-statements-entry-point>", """
    {
      // Code size       44 (0x2c)
      .maxstack  2
      .locals init (System.Span<char> V_0,
                    System.Span<char> V_1)
      IL_0000:  ldstr      "0123"
      IL_0005:  call       "char[] string.ToCharArray()"
      IL_000a:  call       "System.Span<char> System.MemoryExtensions.AsSpan<char>(char[])"
      IL_000f:  stloc.0
      IL_0010:  ldloca.s   V_0
      IL_0012:  ldc.i4.1
      IL_0013:  call       "System.Span<char> System.Span<char>.Slice(int)"
      IL_0018:  stloc.1
      IL_0019:  ldloca.s   V_1
      IL_001b:  constrained. "System.Span<char>"
      IL_0021:  callvirt   "string object.ToString()"
      IL_0026:  call       "void System.Console.Write(string)"
      IL_002b:  ret
    }
    """);
}
```

Failure of `CompileAndVerify`:

```
System.Exception : IL Verify failed unexpectedly:
[<<Main>$>g__SliceSpan|0_0]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x8 }
```

Reply:
Was this a response to https://github.com/dotnet/roslyn/pull/74643#discussion_r1707599978? First, look at other usages of `ExecutionConditionUtil`: the whole point I'm making here is that you shouldn't skip the test on non-CoreCLR. Instead, you should just avoid executing the compiled binary.

As to the error, you'd have to actually show the IL of the local function that's causing it, but ILVerify is not happy with something about it.

Reply:
> Was this a response to https://github.com/dotnet/roslyn/pull/74643#discussion_r1707599978?

No. I thought `if (!ExecutionConditionUtil.IsCoreClr)` were correct. It's my fault that "Do you know how to fix?" is too ambiguous.

> avoid executing the compiled binary.

Do you mean we have to skip only "Verify" (but still run "Compile") in net472? Do we have only to `CreateCompilation(WithIndexAndRange)` in net472?

> show the IL of the local function that's causing it,

I'm afraid I have no idea about how to do it.

FYI, I tried to make net472 able to compile `Span` by modifying `TestSources` but the exception got worse. This is why I don't want to run these tests in net472. Only `SingleOverloadString` was able to be run even in net472 thanks to `CompileAndVerifyWithIndexAndRange`.

```cs
var comp = CreateCompilationWithIndexAndRange(
    new[] { source, TestSources.GetSubArray, TestSources.Span, TestSources.MemoryExtensions },
    TestOptions.UnsafeReleaseExe);

var verify = CompileAndVerify(comp, expectedOutput: "123");
```

```diff
--- a/src/Compilers/Test/Utilities/CSharp/TestSources.cs
+++ b/src/Compilers/Test/Utilities/CSharp/TestSources.cs
@@ -85,6 +85,16 @@ public ref T Current
         public static implicit operator ReadOnlySpan<T>(Span<T> span) => new ReadOnlySpan<T>(span.arr);

         public Span<T> Slice(int offset, int length) => new Span<T>(this.arr, offset, length);
+
+        public override string ToString()
+        {
+            if (typeof(T) == typeof(char))
+            {
+                return new string((char[])(object)this.arr);
+            }
+
+            return ""System.Span<"" + typeof(T).Name + "">["" + Length + ""]"";
+        }
     }

     public readonly ref struct ReadOnlySpan<T>
@@ -162,6 +172,16 @@ public ref readonly T Current
         public static implicit operator ReadOnlySpan<T>(string stringValue) => string.IsNullOrEmpty(stringValue) ? default : new ReadOnlySpan<T>((T[])(object)stringValue.ToCharArray());

         public ReadOnlySpan<T> Slice(int offset, int length) => new ReadOnlySpan<T>(this.arr, offset, length);
+
+        public override string ToString()
+        {
+            if (typeof(T) == typeof(char))
+            {
+                return new string((char[])(object)this.arr);
+            }
+
+            return ""System.ReadOnlySpan<"" + typeof(T).Name + "">["" + Length + ""]"";
+        }
     }

     public readonly ref struct SpanLike<T>
@@ -523,6 +543,8 @@ public static bool SequenceEqual<T> (this Span<T> span, ReadOnlySpan<T> other) w
         }

         public static ReadOnlySpan<char> AsSpan(this string text) => string.IsNullOrEmpty(text) ? default : new ReadOnlySpan<char>(text.ToCharArray());
+
+        public static Span<T> AsSpan<T>(this T[] array) => new Span<T>(array);
     }
 }";
     }
```

```
  メッセージ: 
System.Exception : IL Verify failed unexpectedly:
[.ctor]: Unmanaged pointers are not a verifiable type. { Offset = 0x1 }
[GetEnumerator]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0xb }
[op_Implicit]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x6 }
[op_Implicit]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0xb }
[Slice]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0xd }
[get_Item]: Unexpected type on the stack. { Offset = 0xe, Found = readonly address of 'T', Expected = address of 'T' }
[.ctor]: Unmanaged pointers are not a verifiable type. { Offset = 0x1 }
[GetEnumerator]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0xb }
[op_Implicit]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x9 }
[op_Implicit]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x13 }
[op_Implicit]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x18 }
[op_Implicit]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x22 }
[Slice]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0xd }
[ToArray]: Unmanaged pointers are not a verifiable type. { Offset = 0x0 }
[AsSpan]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x13 }
[AsSpan]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x1d }
[AsSpan]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x6 }

  スタック トレース: 
CompilationVerifier.ILVerify(Verification verification) 行 377
CompilationVerifier.Emit(String expectedOutput, Boolean trimOutput, Nullable`1 expectedReturnCode, String[] args, IEnumerable`1 manifestResources, EmitOptions emitOptions, Verification peVerify, SignatureDescription[] expectedSignatures) 行 276
CommonTestBase.Emit(Compilation compilation, IEnumerable`1 dependencies, IEnumerable`1 manifestResources, SignatureDescription[] expectedSignatures, String expectedOutput, Boolean trimOutput, Nullable`1 expectedReturnCode, String[] args, Action`1 assemblyValidator, Action`1 symbolValidator, EmitOptions emitOptions, Verification verify) 行 191
CommonTestBase.CompileAndVerifyCommon(Compilation compilation, IEnumerable`1 manifestResources, IEnumerable`1 dependencies, Action`1 sourceSymbolValidator, Action`1 assemblyValidator, Action`1 symbolValidator, SignatureDescription[] expectedSignatures, String expectedOutput, Boolean trimOutput, Nullable`1 expectedReturnCode, String[] args, EmitOptions emitOptions, Verification verify) 行 103
CSharpTestBase.CompileAndVerify(Compilation compilation, IEnumerable`1 manifestResources, IEnumerable`1 dependencies, Action`1 sourceSymbolValidator, Action`1 validator, Action`1 symbolValidator, SignatureDescription[] expectedSignatures, String expectedOutput, Boolean trimOutput, Nullable`1 expectedReturnCode, String[] args, EmitOptions emitOptions, Verification verify) 行 1048
IndexAndRangeTests.SingleOverloadSpan() 行 4284
RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

Reply:
It does not surprise me that it fails on net472; there's a `verification` or similar parameter to `CompileAndVerify` that controls what we do with verification failures. I think we still need to understand the non472 failure. The message contained the location of the failure: `<<Main>$>g__SliceSpan|0_0`. That's the local function you defined.

Reply:
I managed to suppress the exception by passing `Verification.Skipped` as a last resort.